src/dfx/src/lib/models/canister.rs

```diff
@@ -197,10 +210,31 @@ impl Canister {
             remove_metadata(&mut m, name);
 
             add_metadata(&mut m, visibility, name, data);
+            modified = true;
+        }
+
+        // If not modified and not set "gzip" explicitly, copy the wasm file directly so that hash match.
+        if !modified && !info.get_gzip() {
+            dfx_core::fs::copy(build_output_wasm_path, &wasm_path)?;
```

Comment:
The deleted logic does following things:
1. copy the wasm file from `build_output.wasm` (file1) to `canister.info.get_build_wasm_path()` (file2) if these two paths are not the same;
2. `apply_metadata` to file2

The new implementation does following instead:
1. read file1 as wasm module
2. do the post-process on the wasm module
3. gzip if need
4. write the wasm (maybe gzipped) to file2

The e2e tests in `build.bash`, `install.bash` and `deploy.bash` make sure that the wasm file are generated at the correct path (file2) so that the canister can be installed.